s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -234,14 +234,19 @@ import scala.util.{Failure, Success, Try}
               Source.fromFuture(entity.discardBytes().future().map(_ => None))
             case HttpResponse(code, _, entity, _) =>
               Source.fromFuture {
-                Unmarshal(entity).to[String].map { err =>
-                  throw new S3Exception(err, code)
-                }
+                unmarshalError(code, entity)
               }
           }
       }
       .mapMaterializedValue(_ => NotUsed)
 
+  private def unmarshalError(code: StatusCode, entity: ResponseEntity)(implicit mat: Materializer) = {
+    import mat.executionContext
+    Unmarshal(entity).to[String].map { err =>
```

Comment:
A bit weird way to use the marshaller but still better than all the duplication from before.